lib/etcd/restore.go

```diff
@@ -97,20 +106,90 @@ func Restore(ctx context.Context, conf RestoreConfig) error {
 			if err != nil {
 				return err
 			}
-		} else {
+		}
+		if !conf.SkipV3 && node.V3 != nil {
 			err = restoreNodeV3(ctx, conf, node.V3, clientv3)
 			if err != nil {
 				return err
 			}
 		}
-
 	}
 
 	conf.Log.Info("Completed etcd restore.")
 
 	return nil
 }
 
+// OfflineRestore restores the etcdv3 database to an offline etcd nodes snapshot DB.
+func OfflineRestore(ctx context.Context, conf RestoreConfig, dir string) error {
+	err := conf.CheckAndSetDefaults()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	conf.Log.Info("Starting etcd offline restore.")
+
+	file, err := os.Open(conf.File)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	decoder := json.NewDecoder(file)
+
+	// The first record in the backup file should be the version
+	var version backupVersion
+	err = decoder.Decode(&version)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if version.Version != FileVersion {
+		return trace.BadParameter("Unsupported backup version %v. ", version.Version)
+	}
+
+	// Open the etcd database directly
+	var be backend.Backend
+	bch := make(chan struct{})
+	go func() {
+		defer close(bch)
+		be = backend.NewDefaultBackend(filepath.Join(dir, "member", "snap", "db"))
+	}()
+	select {
+	case <-bch:
+	case <-time.After(30 * time.Second):
+		return trace.BadParameter("timed out waiting for etcd db lock").AddField("dir", dir)
+	}
+	defer be.Close()
+
+	// Restore the DB
+	for {
+		var node etcdBackupNode
+		err := decoder.Decode(&node)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if node.V3 != nil {
+			writeV3KV(be, node.V3)
```

Comment:
What about V2 records? They still will be restored using the API?

Reply:
correct, v2 records still use the online restore process, and all etcd clients are restarted, because we control and know all the V2 clients. Eventually we'll be able to use the V3 backend for V2 API and drop the V2 stuff, but in etcd 3.4 I don't know that we want to risk it, upstream doesn't advertise using the v3 backend with the v2 API as stable. Likely the next time we touch this, this will be the case, and we won't need to worry about it.